src/java/org/apache/cassandra/service/replication/migration/MigrationRouter.java

```diff
@@ -0,0 +1,307 @@
+        // completed migration to tracked. Therefore, we only need to split ranges when
+        // migrating to tracked replication. For untracked migrations, all reads use untracked.
+        if (!isTracked || migrationInfo == null)
+            return ImmutableList.of(new RangeReadWithReplication(command, isTracked));
+
+        // Get pending ranges for this table
+        NormalizedRanges<Token> tablePendingRanges = migrationInfo.pendingRangesPerTable.get(command.metadata().id());
+
+        // No pending ranges for this table - entire range uses current protocol
+        if (tablePendingRanges == null)
+            return ImmutableList.of(new RangeReadWithReplication(command, isTracked));
+
+        // split into pending (untracked) and non-pending (tracked) ranges
+        List<RangeReadWithReplication> result = splitRangeByPendingRanges(
+            command,
+            command.dataRange().keyRange(),
+            tablePendingRanges,
+            isTracked);
+
+        // Validate the splits
+        validateSplitContiguity(command, result);
+
+        return result;
+    }
+
+    public static boolean shouldUseTrackedForWrites(ClusterMetadata metadata, String keyspace, TableId tableId, Token token)
+    {
+        if (SchemaConstants.isSystemKeyspace(keyspace))
+            return false;
+
+        KeyspaceMigrationInfo migrationInfo = metadata.mutationTrackingMigrationState
+                                              .getKeyspaceInfo(keyspace);
+
+        if (migrationInfo == null)
+            return metadata.schema.getKeyspaceMetadata(keyspace).params.replicationType.isTracked();
+
+        boolean isTracked = metadata.schema.getKeyspaceMetadata(keyspace).params.replicationType.isTracked();
+        return migrationInfo.shouldUseTrackedForWrites(isTracked, tableId, token);
+    }
+
+    public static class RoutedMutations
+    {
+        public final List<? extends IMutation> trackedMutations;
+        public final List<? extends IMutation> untrackedMutations;
+
+        public RoutedMutations(List<? extends IMutation> tracked, List<? extends IMutation> untracked)
+        {
+            this.trackedMutations = tracked;
+            this.untrackedMutations = untracked;
+        }
+    }
+
+    /**
+     * Route a list of mutations, splitting them into tracked and untracked groups.
+     */
+    public static RoutedMutations routeMutations(List<? extends IMutation> mutations)
+    {
+        List<IMutation> tracked = new ArrayList<>();
+        List<IMutation> untracked = new ArrayList<>();
+        ClusterMetadata cm = ClusterMetadata.current();
+
+        for (IMutation mutation : mutations)
+        {
+            IMutation untrackedMutation = mutation.filter(tid -> !shouldUseTrackedForWrites(cm, mutation.getKeyspaceName(), tid, mutation.key().getToken()));
+            if (untrackedMutation != null)
+                untracked.add(mutation);
```

Comment:
This is adding `mutation` not `untrackedMutation`. Seems wrong.

Also... if there are no failing tests because of this bug then a test needs to be added. 100% this particular kind of mistake should trigger a test failure somewhere.